### DISCUSSION TITLE
GetTagList & Global Variables

### DISCUSSION THREAD

**Discussion by badrinarayanans** [dmroeder/pylogix#181] [Q&A] [answered]
Hi, I am using CCW and Micro 850 Simulator and the GetTagList function against this simulator doesn't return all global variables. Especially variables of type TON, variables of type User Defined Function. The simulator also has two programs written, the GetProgramsList() doesn't return any programs. Are these limitations because I am using the simulator? Or is there something I am missing? Thanks.

**Comment by ottowayi** [comment, marked as answer]
I've noticed it will exclude any array tags that aren't 0-based.  It's not pylogix excluding it, when you ask for for the tag list the PLC never includes them in response.  My best guess at why is the Micro 800's were from an acquired company and the CIP standard doesn't support arrays that don't start at 0. So when they added ENIP support they chose to exclude them.

**Reply by badrinarayanans** [reply]
Thanks for the clarification. It helped.

**Reply by ottowayi** [reply]
I haven't tried with those specific tag types, but there might be a similar situation.  I discovered those stupid controllers are full of special cases when developing my own ENIP library.  If you want to compare with mine (pycomm3 on my profile) and verify that it's the PLC and not pylogix you can.  I need to re-setup my environment after moving or else I would have tried it myself.

**Reply by dmroeder** [reply]
@badrinarayanans if you can email me a sample program for the simulator, I can try to test.  Being that you cannot access local variables, it wouldn't surprise me if the program names were unavailable as well, though I've never tested that.  I originally didn't support the Micro800 because I don't use them.  After I found out that it took only small changes to read/write, I decided to make the changes.  As @ottowayi mentioned, they are very limited in what CIP objects they support

**Reply by badrinarayanans** [reply]
> I haven't tried with those specific tag types, but there might be a similar situation. I discovered those stupid controllers are full of special cases when developing my own ENIP library. If you want to compare with mine (pycomm3 on my profile) and verify that it's the PLC and not pylogix you can. I need to re-setup my environment after moving or else I would have tried it myself.

The pycomm3 library also didn't return the variables of type struct/array. So, this is related to the PLC. Thanks.

**Reply by dmroeder** [reply]
Thanks for the update.
